app/controllers/concerns/application_shared.rb

```diff
@@ -92,11 +103,16 @@ def find_default_taxonomy(user, taxonomy)
 
   def find_session_taxonomy(taxonomy, user)
     available = user.send("my_#{taxonomy.pluralize}")
-    determined_taxonomy = available.where(:id => request.session["#{taxonomy}_id"]).first
+    determined_taxonomy = available.where(:id => session["#{taxonomy}_id"]).first
     # warn user if taxonomy stored in session does not exist and delete it from session (probably taxonomy has been deleted meanwhile)
     if determined_taxonomy.nil?
-      warning _("%s you had selected as your context has been deleted") % taxonomy.capitalize unless api_request?
-      request.session["#{taxonomy}_id"] = nil
+      if api_request?
+        not_found _("%{taxonomy} stored in session with id %{id} not found") % { :taxonomy => taxonomy.capitalize, :id => params["#{taxonomy}_id"] }
+        return false
+      else
+        warning _("%s you had selected as your context has been deleted") % taxonomy.capitalize unless api_request?
```

Comment:
The `unless api_request?` is tautology, as we are in the `else` branch of `if api_request?`